### PR TITLE
Precise test cases for encoding rules; some fixes to existing compact uint test cases.

### PR DESCRIPTION
**Pull request by kyegupov**
The codec format needs to be stable and precisely defined so that alternative implementations can be verified against it. An easiest step in that direction is defining some tests that verify the actual binary encoding. I've tried to cover all the most interesting cases.

### CODE REVIEW

**Comment by bkchr** [comment]
Hey, the tests do not compile, can you fix this?

**Comment by kyegupov** [comment]
@bkchr done, sorry

**Comment by bkchr** [comment]
[clabot:check]

**Comment by bkchr** [comment]
@kyegupov no problem :) Ty for the contribution. We just need to wait for the cla bot.

**Comment by kyegupov** [comment]
@bkchr I've signed the CLA just now (maybe the repo README should have a note about it?), do I need to update anything?

**Comment by kyegupov** [comment]
[clabot:check]

**Comment by ddorgan** [comment]
[clabot:check]

File: src/codec.rs [hunk 1]
```diff
@@ -464,6 +464,7 @@ impl<T: Decode, E: Decode> Decode for Result<T, E> {
 }
 
 /// Shim type because we can't do a specialised implementation for `Option<bool>` directly.
+#[derive(Eq, PartialEq, Clone, Copy, Debug)]
```
Comment (rphmeier) [inline src/codec.rs]: The `Debug` derive should be on `std` only

**Comment by rphmeier** [comment]
LGTM except for nit. Thanks

**Comment by kyegupov** [comment]
@rphmeier fixed

**Pull request merged**